### PR TITLE
Remove dependency on custom Prometheus fork

### PR DESCRIPTION
**Pull request by juliusv**
Sorry for the huge PR - I think this wasn't really feasible to split up into multiple, still compiling and working PRs. When changing the origin repo of the Prometheus vendoring to point to `prometheus/prometheus` vs. `tomwilkie/prometheus`, that requires basically touching and fixing up everything at once.

Things in this PR:

- Changes the vendoring of Prometheus to use the official upstream master.
- The static web (graph UI) assets were partially in our custom fork, partially in upstream (but not in a reusable fashion). I copied the necessary assets, as well as added build steps to build the compiled-in static assets in `bindata.go`.
- The web API copy is now deleted. Instead, we use the new context support in the upstream WebUI->PromQL->Storage stack. Instead of building a new MergeQuerier for every request, creating a PromQL engine from that, and handing it to the web UI, we can now simply add the user info to the request context at the very beginning and let the web UI and PromQL hand the context all the way down to our custom storage, which extracts that info from the storage again.
- Use the opened-up upstream chunk interfaces from the ingester.
- Since there was no agreement upstream to open up the fingerprint mapper, fingerprint locker, and memory series code to outside users, I had to copy over the code for them. I simplified them as far as I could: removed persistence support and deleted all unused methods. Further simplifications are possible in the future.
- A number of other adjustments that were necessary.

I added three `TODO(juliusv)` markers that need further upstream support to be fixed again:

* The upstream chunks currently don't have a `utilization` method and there's no way to get to this info anymore, so I disabled the instrumentation for it for now - high priority to enable this again, as this histogram also functions as a counter of stored chunks:
```go
// TODO(juliusv): Re-enable upstream support for this.
// i.chunkUtilization.Observe(chunk.C.utilization())
```

* When decoding a chunk in the querier, we can't get the number of samples in the chunk before decoding it anymore, leading to a probably negligible performance overhead:

```go
// TODO(juliusv): Pre-allocate this with the right length again once we
// add a method upstream to get the number of samples in a chunk.
var samples []model.SamplePair
```

* The upstream web API expects a `local.Storage` as the second constructor argument. I extended our `MergeQuerier` to satisfy this interface, although most of the methods are not actually called. Plan: make the expected interface in the upstream web API narrower to only ask for the methods the web API is actually using.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -159,55 +158,51 @@ func QueryHandler(querier Querier) http.Handler {
...
 			return
 		}
 
-		resp := &generic.GenericReadResponse{}
+		resp := &ReadResponse{}
 		for _, ss := range res {
-			ts := &generic.TimeSeries{
-				Name: proto.String(string(ss.Metric[model.MetricNameLabel])),
-			}
+			ts := &TimeSeries{}
```
Comment (jml) [inline server.go]: Name no longer needed?

File: ingester/ingester.go [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+			index:      newInvertedIndex(),
+		}
+		state.mapper = newFPMapper(state.fpToSeries)
+		i.userState[userID] = state
+	}
+	return state, nil
+}
+
+// NeedsThrottling implements storage.SampleAppender.
+func (*Ingester) NeedsThrottling(_ context.Context) bool {
```
Comment (jml) [inline ingester/ingester.go]: Is this something we intend to always leave as `false`?

File: ingester/ingester.go [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+	if err != nil {
+		return err
+	}
+
+	fp, series, err := state.getOrCreateSeries(sample.Metric)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		state.fpLocker.Unlock(fp)
```
Comment (jml) [inline ingester/ingester.go]: What locks this such that it needs an unlock?

File: ingester/ingester.go [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+	}()
+
+	if sample.Timestamp == series.lastTime {
+		// Don't report "no-op appends", i.e. where timestamp and sample
+		// value are the same as for the last append, as they are a
+		// common occurrence when using client-side timestamps
+		// (e.g. Pushgateway or federation).
+		if sample.Timestamp == series.lastTime &&
```
Comment (jml) [inline ingester/ingester.go]: This first conditional is unnecessary, as it's guaranteed by the prior one.

File: ingester/ingester.go (hunk 3 above)
Comment (jml) [inline ingester/ingester.go]: Oh I see, `getOrCreateSeries` does. I think that's probably unusual enough that a comment would be helpful.

File: ingester/ingester.go [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+	for _, fp := range fps {
+		state.fpLocker.Lock(fp)
+		series, ok := state.fpToSeries.get(fp)
+		if !ok {
+			state.fpLocker.Unlock(fp)
+			continue
+		}
+
+		values, err := samplesForRange(series, from, through)
+		state.fpLocker.Unlock(fp)
```
Comment (jml) [inline ingester/ingester.go]: I wonder if it's worth extracting this logic into a function just so the `Unlock` can be put in a `defer`.

File: ingester/ingester.go [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+
+func (i *Ingester) flushUser(userID string, immediate bool) {
+	log.Infof("Flushing user %s...", userID)
+	defer log.Infof("Done flushing user %s.", userID)
+
+	i.userStateLock.Lock()
+	userState, ok := i.userState[userID]
+	i.userStateLock.Unlock()
+
+	// This should happen, right?
```
Comment (jml) [inline ingester/ingester.go]: I think this should say "This should not happen", but maybe I'm wrong.

File: ingester/ingester.go [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+	if !ok {
+		return
```
Comment (jml) [inline ingester/ingester.go]: This seems like exactly the case for a `panic`: there's a non-recoverable programming error.

**Comment by jml** [comment]
Vendoring commit also looks ok.

File: server.go (hunk 1 above)
Comment (juliusv) [inline server.go]: The metric name is just another label now in the protobuf (both upstream in the generic write protobufs and in our generic read query ones in Prism).

**Comment by juliusv** [comment]
@jml Thank you!

Indeed, some comments to look at there in the ingester, but they all concern code that was just copied as is. So I'd prefer addressing them in a followup.

**Comment by jml** [comment]
Great. I don't think I've found anything here that needs to be addressed here, so please merge when ready.

**Pull request merged**